### ISSUE TITLE
Better use of Make + dockerized dev environment

### ISSUE THREAD

**Issue by customcommander** [ramda/ramda#3034]
I can see there's a Makefile but I can't see where Make is actually invoked anywhere. Is this right?

Is there an appetite to reconsider using Make? (I'm a big fan of Make)

The current Makefile doesn't leverage the _targets - prerequisites_ model as well as it could. It's possible to only (re)build what's needed and it's not a very complex thing to do with Make.

The development environment could also be dockerized so as to make sure everybody (on any platform) develops against the same version of Node, with Make and the same shell. This would get rid of packages such as cross-env and rimraf.

If there's even a mild interest I can take a stab at it and show what's possible.

**Comment by CrossEye** [comment]
I'm mixed about this.

I would love to have our build as simple as possible.  Requiring Docker in order to build seems a bit much.  I also want to make sure that Windows users aren't left in the cold.  But I don't see any good way to achieve both a simple build and Docker-free Windows support.

For now, though, `make` is how we create and publish our releases, so I guess we're already away from the Windows world.  (I know I've actually done it on Ubuntu by calling XYZ directly, though.

----------

In other words, my thoughts on this are scrambled!

**Comment by customcommander** [comment]
That's fine. I'll have a go and we'll see where that goes.

**Comment by hitmands** [comment]
I'd suggest to keep the dev life cycle within the NPM scripts, 
this would guarantee we are cross platform and would also let us decommission tools like make, etc.

Ramda should work as any regular npm project...

```
npm install
npm run lint
npm run compile
npm publish
```

I don't think we really need any other superstructure, 
we should also keep the CI in mind...

any job that you can run locally should also be runnable on github actions.

**Comment by customcommander** [comment]
I don't think we can assume that all NPM projects are built and managed the same way ;)

NPM scripts are indeed very popular as Make is. Both are well known, documented solutions. We should use whatever works for the team while remaining open to alternatives.

I was just trying to propose a reproducible environment that can work the same way both on a developer machine _and_ on a CI/CD setup. This is what I was aiming for.

Make does indeed add a dependency but combined with Docker it would also remove other stuff that exist only for addressing the issue of reproducible building environment. I'm thinking of `cross-env` and `rimraf`.

Now I'm not particularly fussed. If there's no interest in this I'm totally fine with it :-)

**Comment by hitmands** [comment]
I think a discussion around "how we can improve the tooling" is needed here, 
I wouldn't really close this issue... some of the building tools don't necessarily reflect the state of the art.

It's just a matter of finding the way to go!

**Comment by customcommander** [comment]
Sure when you know what better means I'll be happy to chip in.
